### ISSUE TITLE
Reduce size of syntax::tokenstream::TokenTree

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust#54535]
#53902 raised the size of TokenTree from 32 bytes to 40 bytes, resulting in increased max-rss across a variety of benchmarks by up to 6%.

Here are the data structures involved: [**playground**](https://play.rust-lang.org/?gist=7ccf1f3d1fee4aeeacd70d1047da6981&version=stable&mode=debug&edition=2015)

Even among the 32 bytes a lot of those were padding. It should be possible to rearrange a few things to drop back to 32 bytes while retaining all the information currently represented.

@nnethercote

**Comment by eddyb** [comment]
@dtolnay Were there APIs on `Delimited`? I think that's a reasonable inline.
Would it also be possible to move `DelimSpan` inside `Delimited`? Or would that not help?

**Comment by dtolnay** [comment]
Inlining DelimSpan does not help.

There are five methods on Delimited:

https://github.com/rust-lang/rust/blob/a072d1bca6282d1d2fbec103d12393527eb5a868/src/libsyntax/tokenstream.rs#L46-L81

**Comment by eddyb** [comment]
How does `DelimSpan` compare to what `{open,close}_tt` compute?
Also, `{open,close}_token` could be methods on `DelimToken` itself.
So overall, losing `Delimited` doesn't feel too bad.

**Comment by dtolnay** [comment]
The `{open,close}_tt` methods are passed in a Span. DelimSpan is where the span passed in comes from.

**Comment by whitfin** [comment]
@dtolnay I was going to take a look at this but it seems it might have already been done (and so this can be closed). The current `TokenTree::Delimited` looks like this:

```rust
pub enum TokenTree {
    Token(Span, token::Token),
    Delimited(DelimSpan, DelimToken, TokenStream),
}
```

**Comment by dtolnay** [comment]
Thanks, looks like this was fixed in #56369.
